src/psyclone/gocean1p0.py

```diff
@@ -173,11 +175,24 @@ class GOInvoke(Invoke):
         method so that we generate GOcean specific invocation code and
         provides three methods which separate arguments that are arrays from
         arguments that are {integer, real} scalars. '''
+
     def __init__(self, alg_invocation, idx):
+        '''Constructor for the GOcean-specific invoke class.
+        :param alg_invocation: Node in the AST describing the invoke call.
+        :type alg_invocation: :py:class:`psyclone.parse.InvokeCall`
+        :param int idx: The position of the invoke in the list of invokes
+                        contained in the Algorithm.
+        '''
+
         if False:  # pylint: disable=using-constant-test
             self._schedule = GOSchedule(None)  # for pyreverse
         Invoke.__init__(self, alg_invocation, idx, GOSchedule)
 
+        # Add profiling nodes to schedule if automatic profiling has been
+        # requested.
+        from psyclone.profiler import Profiler
+        Profiler.add_profile_nodes(self.schedule, GOLoop)
```

Comment:
I've given this some more thought as it seems to me that automatic profiling shouldn't require API-specific code. In fact, the way you've written `add_profile_nodes` already means that it will work for all APIs if we pass `Loop` instead of the API-specific sub-class. However, we can only instrument a Schedule once it is complete and that part *is* API specific. Therefore, how about moving the code to add profiling into the constructor of the `Invokes` base class? e.g.:

        for idx, alg_invocation in enumerate(alg_calls.values()):
            my_invoke = Invoke(alg_invocation, idx)
            # Automatically instrument this Invoke for profiling (if enabled)
            Profiler.add_profile_nodes(my_invoke.schedule, Loop)

Reply:
I quickly tested it, and it seems to work - I somehow dismissed the Invokes base class too quickly ;)

But I also noticed that my regular expression tests probably don't detect duplicated ProfileStart etc. calls (i.e. your new code AND my old code together still passes all tests ... which is strange).
I'll test this and add the other mentioned issues as well.